kubernetes/client/cache.go

```diff
@@ -0,0 +1,164 @@
+	AddEventHandler(handler EventHandler)
+}
+
+// Lister is an interface for things that can list objects for all namespaces or by namespace
+type Lister interface {
+	List() []interface{}
+	ByNamespace() []interface{}
+}
+
+// Cache is an implementation of a List/Watch cache
+type Cache struct {
+	Extractor        func(interface{}) *V1ObjectMeta
+	Lister           ObjectLister
+	Watcher          ObjectWatcher
+	allObjects       []objEntry
+	namespaceObjects map[string][]objEntry
+	eventHandlers    []EventHandler
+}
+
+func (c *Cache) AddEventHandler(handler EventHandler) {
+	c.eventHandlers = append(c.eventHandlers, handler)
+}
+
+func (c *Cache) Run() {
+	for {
+		if err := c.ListWatch(); err != nil {
+			log.Printf("%s\n", err.Error())
+		}
+	}
+}
+
+func (c *Cache) ListWatch() error {
+	objects, resourceVersion, err := c.Lister()
+	if err != nil {
+		return err
+	}
+	for ix := range objects {
+		meta := c.Extractor(objects[ix])
+		c.AddOrUpdate(meta, objects[ix])
+	}
+	results, errors := c.Watcher(resourceVersion)
+	for {
+		select {
+		case result, ok := <-results:
+			if !ok {
+				return nil
+			}
+			c.ProcessResult(result)
+		case err := <-errors:
+			return err
+		}
+	}
+}
+
+func (c *Cache) ProcessResult(res *Result) {
+	metadata := c.Extractor(res.Object)
+	if res.Type == "ADDED" || res.Type == "MODIFIED" {
+		c.AddOrUpdate(metadata, res.Object)
+		return
+	} else if res.Type == "DELETED" {
+		c.Delete(metadata, res.Object)
+	}
+}
+
+func (c *Cache) AddOrUpdate(metadata *V1ObjectMeta, obj interface{}) {
+	var oldObj interface{}
+	c.allObjects, oldObj = InsertOrUpdate(c.allObjects, metadata, obj)
+	if len(metadata.Namespace) > 0 {
+		c.namespaceObjects[metadata.Namespace], _ =
+			InsertOrUpdate(c.namespaceObjects[metadata.Namespace], metadata, obj)
+	}
+	for ix := range c.eventHandlers {
+		if oldObj == nil {
+			c.eventHandlers[ix].OnAdd(obj)
+		} else {
+			c.eventHandlers[ix].OnUpdate(oldObj, obj)
+		}
+	}
+}
+
+func (c *Cache) Delete(metadata *V1ObjectMeta, obj interface{}) {
+	var deleted bool
+	c.allObjects, deleted = Delete(c.allObjects, metadata)
+	if len(metadata.Namespace) > 0 {
+		c.namespaceObjects[metadata.Namespace], _ =
+			Delete(c.namespaceObjects[metadata.Namespace], metadata)
+	}
+	if deleted {
+		for ix := range c.eventHandlers {
+			c.eventHandlers[ix].OnDelete(obj)
+		}
+	}
+}
+
+func (c *Cache) List() []interface{} {
+	result := make([]interface{}, len(c.allObjects))
+	for ix := range c.allObjects {
+		result[ix] = c.allObjects[ix].obj
+	}
+	return result
+}
+
+func (c *Cache) ByNamespace(namespace string) []interface{} {
+	list := c.namespaceObjects[namespace]
+	result := make([]interface{}, len(list))
+	for ix := range list {
+		result[ix] = list[ix].obj
+	}
+	return result
+}
+
+func InsertOrUpdate(list []objEntry, metadata *V1ObjectMeta, obj interface{}) ([]objEntry, interface{}) {
+	ix := FindObject(list, metadata)
+	if ix == -1 {
+		return append(list, objEntry{metadata: metadata, obj: obj}), nil
+	} else {
```

Comment:
no need for 'else' as there is 'return' above

Reply:
done.